packages/keyring/src/keyring.ts

```diff
@@ -0,0 +1,319 @@
+import blake from 'blakejs'
+import { HDKey } from 'ethereum-cryptography/hdkey'
+import { bytesToHex } from 'ethereum-cryptography/utils'
+
+import { decryptMnemonic, MnemonicLength, mnemonicStringToUint8Array } from '@/mnemonic'
+
+export type NonSensitiveAddressData = {
+  hash: AddressHash
+  index: number
+  publicKey: string
+}
+
+type SensitiveAddressData = NonSensitiveAddressData & {
+  privateKey: Uint8Array
+}
+
+type GenerateAddressProps = {
+  group?: number
+  addressIndex?: number
+  skipAddressIndexes?: number[]
+}
+
+class Keyring {
+  private hdPath = "m/44'/1234'/0'/0"
+  private hdWallet: HDKey | null
+  private root: HDKey | null
+  private addresses: SensitiveAddressData[]
+
+  constructor() {
+    this.addresses = []
+    this.hdWallet = null
+    this.root = null
+  }
+
+  // PUBLIC METHODS
+
+  public clearCachedSecrets = () => {
+    this.addresses.map((address) => resetArray(address.privateKey))
+    this.addresses = []
+    this.hdWallet = null
+    this.root = null
+  }
+
+  public generateRandomMnemonic = (mnemonicLength: MnemonicLength = 24): Uint8Array => {
+    const strength = mnemonicLength === 24 ? 256 : 128
+    const mnemonic = metamaskBip39.generateMnemonic(wordlist, strength)
+
+    this._initFromMnemonic(mnemonic, '')
+
+    return mnemonic
+  }
+
+  public importMnemonicString = (mnemonicStr: string): Uint8Array => {
+    if (!mnemonicStr) throw new Error('Keyring: Cannot import mnemonic, mnemonic not provided')
+
+    if (!metamaskBip39.validateMnemonic(mnemonicStr, wordlist))
+      throw new Error('Keyring: Cannot import mnemonic, invalid mnemonic provided')
+
+    console.log('Mnemonic leaked to memory as a string while importing wallet. This is expected.')
+
+    const mnemonic = mnemonicStringToUint8Array(mnemonicStr)
+
+    this.clearCachedSecrets()
+    this._initFromMnemonic(mnemonic, '')
+
+    return mnemonic
+  }
+
+  public initFromEncryptedMnemonic = (encryptedMnemonic: string, password: string, passphrase: string) => {
+    const { version, decryptedMnemonic } = decryptMnemonic(encryptedMnemonic, password)
+
+    this.clearCachedSecrets()
+    this._initFromMnemonic(decryptedMnemonic, passphrase)
+
+    encryptedMnemonic = ''
+    password = ''
+    passphrase = ''
+    resetArray(decryptedMnemonic)
+
+    return version
+  }
+
+  public generateAndCacheAddress = (props: GenerateAddressProps): NonSensitiveAddressData =>
+    this._getNonSensitiveAddressData(this._generateAndCacheAddress(props))
+
+  public signTransaction = (txId: string, addressHash: AddressHash): string =>
+    transactionSign(txId, this.exportPrivateKeyOfAddress(addressHash))
+
+  public signMessage = (message: string, addressHash: AddressHash): string =>
+    sign(message, this.exportPrivateKeyOfAddress(addressHash))
+
+  public exportPrivateKeyOfAddress = (addressHash: AddressHash): string =>
+    bytesToHex(this._getAddress(addressHash).privateKey)
+
+  public discoverAndCacheActiveAddresses = async (
```

Comment:
Linked to my previous comment about (not) calling endpoints in the keyring, maybe this method should be moved to somewhere else. But I see why you did this, it feels more confined and handy to have this available in the keyring package...
Is Metamask doing something similar?

Reply:
I did not see any address discovery method in the keyring in Metamask.

Reply:
Addressed in 1ce954a11 in PR https://github.com/alephium/alephium-frontend/pull/504